pilot/pkg/model/sidecar.go

```diff
@@ -0,0 +1,324 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"strings"
+
+	networking "istio.io/api/networking/v1alpha3"
+)
+
+const (
+	wildcardNamespace = "*"
+	wildcardService   = Hostname("*")
+)
+
+// SidecarScope is a wrapper over the Sidecar resource with some
+// preprocessed data to determine the list of services, virtualServices,
+// and destinationRules that are accessible to a given sidecar.
+type SidecarScope struct {
+	// The crd itself. Can be nil if we are constructing the default
+	// sidecar scope
+	Config *Config
+
+	// set of egress listeners, and their associated services.
+	// A sidecar scope should have either ingress/egress listeners or both.
+	EgressListeners []*IstioEgressListenerWrapper
+
+	// // set of ingress listeners
+	// // A sidecar scope should have either ingress/egress listeners or both.
+	// IngressListeners []*IstioIngressListenerWrapper
+
+	// Union of services imported across all egress listeners for use by CDS code.
+	// Right now, we include all the ports in these services.
+	// TODO: Trim the ports in the services to only those referred to by the
+	// egress listeners.
+	allImportedServices []*Service
+
+	// Destination rules imported across all egress listeners
+	allImportedDestinationRules map[Hostname]*Config
+
+	// VirtualServices imported across all egress listeners
+	allImportedVirtualServices []Config
+}
+
+// IstioEgressListenerWrapper is a wrapper for networking.IstioEgressListener object.
+// It has the parsed form of the hosts field, encompassing a list of services
+// per sidecar listener object.
+type IstioEgressListenerWrapper struct {
+	// The actual IstioEgressListener api object from the Config. It can be nil if
+	// this is for the default sidecar scope.
+	IstioListener *networking.IstioEgressListener
+
+	// TODO: Unix domain socket
+
+	// optional. The port on which this listener should be configured. If
+	// omitted, we infer from services imported
+	ListenerPort *Port
+
+	// parsed form of the hosts field
+	// map of namespace and services
+	importMap map[string]Hostname
+
+	// List of services imported by this egress listener
+	// extracted from the importMap above.
```

Comment:
Shouldn't this be a map ( by hostname ) for faster lookup ?

Reply:
(we can optimize later, resolving for now)

Reply:
not everything is a map. When you construct the listeners, you need a time ordered array of services so that you can resolve conflicts in a proper fashion. And thats exactly what this is for.